cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -298,13 +296,45 @@ public void performRunLevelStop(HttpRequest request, HttpResponder responder,
                                   @PathParam("app-id") String appId,
                                   @PathParam("program-type") String type,
                                   @PathParam("program-id") String programId,
-                                  @PathParam("run-id") String runId) throws Exception {
+                                  @PathParam("run-id") String runId,
+                                  @QueryParam("graceful") String gracefulShutdownSecs) throws Exception {
     ProgramType programType = getProgramType(type);
     ProgramId program = new ProgramId(namespaceId, appId, programType, programId);
-    lifecycleService.stop(program, runId);
+    int gracefulShutdownSecsInt;
+    try {
+      gracefulShutdownSecsInt = validateAndGetGracefulShutdownSecsInt(gracefulShutdownSecs);
+    } catch (IllegalArgumentException e) {
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+      return;
+    }
+    try {
+      RunIds.fromString(runId);
```

Comment:
Is there a way to also validate program type and program id?

Reply:
it's validated when creating ProgramType and ProgramId objects